File: pathSumII.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
    	if root is None:
    		return []
        return self.helper(root, [], [], sum, 0)

    def helper(self, root, path, result, target, sum):
    	if root is None:
    		return []

    	if root.left is None and root.right is None:
    		if sum + root.val == target:
    			path.append(root.val)
    			result.append(path)
    			return result

    	if root.left:
    		temp = path[:]
    		temp.append(root.val)
    		self.helper(root.left, temp, result, target, sum + root.val)
    	if root.right:
    		temp = path[:]
    		temp.append(root.val)
    		self.helper(root.right, temp, result, target, sum + root.val)
    	return result

File: JAVA/pathSumII.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        ArrayList<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> temp;
        List<List<Integer>> rest;
        
        if (root == null)
            return result;
        if (root.left == null && root.right == null && root.val == sum){
            temp = new ArrayList<Integer>();
            temp.add(root.val);
            result.add(temp);
        }
        if (root.left != null){
            rest = pathSum(root.left, sum - root.val);
            for (List<Integer> lst: rest)
                lst.add(0, root.val);
            result.addAll(rest);
        }
        if (root.right != null){
            rest = pathSum(root.right, sum - root.val);
            for (List<Integer> lst: rest)
                lst.add(0, root.val);
            result.addAll(rest);
        }
        return result;
    }
}